pkg/loadbalancer/experimental/healthserver.go

```diff
@@ -0,0 +1,296 @@
+				},
+				FrontendParams{
+					Address: lb.L3n4Addr{
+						AddrCluster: fe.Address.AddrCluster,
+						L4Addr: lb.L4Addr{
+							Protocol: lb.TCP,
+							Port:     port,
+						},
+						Scope: lb.ScopeExternal,
+					},
+					Type:        fe.Type,
+					ServiceName: name,
+					ServicePort: port,
+				},
+			)
+			s.params.Writer.UpsertBackends(
+				wtxn,
+				healthServiceName,
+				source.Local,
+				BackendParams{
+					L3n4Addr: lb.L3n4Addr{
+						AddrCluster: healthServerAddr,
+						L4Addr: lb.L4Addr{
+							Protocol: lb.TCP,
+							Port:     port,
+						},
+						Scope: lb.ScopeInternal,
+					},
+					NodeName: s.nodeName,
+					State:    lb.BackendStateActive,
+				},
+			)
+			wtxn.Commit()
+
+		}
+		health.OK(fmt.Sprintf("%d health servers running", len(s.serverByPort)))
+
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-watch:
+		}
+	}
+}
+
+func isExtLocal(fe *Frontend) bool {
+	switch fe.Type {
+	case lb.SVCTypeNodePort, lb.SVCTypeLoadBalancer, lb.SVCTypeExternalIPs:
+		return fe.service.ExtTrafficPolicy == lb.SVCTrafficPolicyLocal
+	default:
+		return false
+	}
+}
+
+func (s *healthServer) cleanupListeners(ctx context.Context) {
+	for _, srv := range s.serverByPort {
+		srv.shutdown(ctx)
+	}
+}
+
+func (s *healthServer) addListener(name lb.ServiceName, port uint16) *httpHealthServer {
+	srv := &httpHealthServer{
+		nodeName: s.nodeName,
+		name:     name,
+		db:       s.params.DB,
+		backends: s.params.Backends,
+	}
+	srv.Server = http.Server{
+		Addr:    fmt.Sprintf("%s:%d", healthServerAddr.Addr().String(), port),
+		Handler: srv,
+	}
+	s.params.Jobs.Add(
+		job.OneShot(
+			fmt.Sprintf("listener-%d", port),
+			func(ctx context.Context, health cell.Health) error {
+				if err := srv.ListenAndServe(); !errors.Is(err, http.ErrServerClosed) {
+					return err
+				}
+				return nil
+			},
+			job.WithRetry(-1, &job.ExponentialBackoff{
+				Min: 200 * time.Millisecond,
+				Max: 10 * time.Second,
+			}),
+		),
+	)
+
+	return srv
+}
+
+func (s *healthServer) removeListener(ctx context.Context, port uint16) {
+	if srv, ok := s.serverByPort[port]; ok {
+		srv.shutdown(ctx)
+		delete(s.serverByPort, port)
+	}
+}
+
+type serviceName struct {
+	Namespace string `json:"namespace"`
+	Name      string `json:"name"`
+}
+
+// healthResponse represents the object returned by the health server
+type healthResponse struct {
+	Service        serviceName `json:"service"`
+	LocalEndpoints int         `json:"localEndpoints"`
+}
+
+type httpHealthServer struct {
+	http.Server
+
+	nodeName string
+	name     lb.ServiceName
+	db       *statedb.DB
+	backends statedb.Table[*Backend]
+}
+
+func (h *httpHealthServer) getLocalEndpointCount() int {
+	txn := h.db.ReadTxn()
+
+	activeCount := 0
+	for be := range h.backends.List(txn, BackendByServiceName(h.name)) {
+		if len(be.NodeName) > 0 && be.NodeName != h.nodeName {
+			continue
+		}
+		if be.State == lb.BackendStateActive {
+			activeCount++
+		}
```

Comment:
looking at the old `upsertService()` code - did we lose the special-casing for L7 services here? It's always reporting 1 active backend (representing the envoy instance).

Reply:
added.